File: java-gpx/src/main/java-gen/com/topografix/gpx/Gpx.java
package com.topografix.gpx;

import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlProperty;
import javax.validation.constraints.NotNull;
import javax.validation.Valid;
import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlElementWrapper;
import com.fasterxml.jackson.annotation.JsonAnySetter;
import com.fasterxml.jackson.annotation.JsonAnyGetter;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

@JsonIgnoreProperties(value = {"schemaLocation"})
public class Gpx
{
    @JacksonXmlProperty(isAttribute = true)
    @NotNull
    private final String version = "1.1";
    @JacksonXmlProperty(isAttribute = true)
    @NotNull
    private String creator;
    @Valid
    private Metadata metadata;
    @JacksonXmlElementWrapper(localName="wpts", useWrapping=false)
    @JacksonXmlProperty(localName="wpt")
    @Valid
    private java.util.List<Wpt> wpts;
    @JacksonXmlElementWrapper(localName="rtes", useWrapping=false)
    @JacksonXmlProperty(localName="rte")
    @Valid
    private java.util.List<Rte> rtes;
    @JacksonXmlElementWrapper(localName="trks", useWrapping=false)
    @JacksonXmlProperty(localName="trk")
    @Valid
    private java.util.List<Trk> trks;
    private java.util.Map<String,Object> extensions = new java.util.TreeMap<>();

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder("Gpx[");
        sb.append("version=").append(version);
        sb.append(", creator=").append(creator);
        sb.append(", metadata=").append(metadata);
        sb.append(", wpts=").append(wpts);
        sb.append(", rtes=").append(rtes);
        sb.append(", trks=").append(trks);
        sb.append(", extensions=").append(extensions);
        return sb.append("]").toString();
    }
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Gpx other = (Gpx)o;
        if (version != null ? !version.equals(other.version) : other.version != null) return false;
        if (creator != null ? !creator.equals(other.creator) : other.creator != null) return false;
        if (metadata != null ? !metadata.equals(other.metadata) : other.metadata != null) return false;
        if (wpts != null ? !wpts.equals(other.wpts) : other.wpts != null) return false;
        if (rtes != null ? !rtes.equals(other.rtes) : other.rtes != null) return false;
        if (trks != null ? !trks.equals(other.trks) : other.trks != null) return false;
        if (extensions != null ? !extensions.equals(other.extensions) : other.extensions != null) return false;
        return true;
    }
    @Override
    public int hashCode() {
        int result = 1;
        result = 31 * result + (version != null ? version.hashCode() : 0);
        result = 31 * result + (creator != null ? creator.hashCode() : 0);
        result = 31 * result + (metadata != null ? metadata.hashCode() : 0);
        result = 31 * result + (wpts != null ? wpts.hashCode() : 0);
        result = 31 * result + (rtes != null ? rtes.hashCode() : 0);
        result = 31 * result + (trks != null ? trks.hashCode() : 0);
        result = 31 * result + (extensions != null ? extensions.hashCode() : 0);
        return result;
    }
    public String getVersion() {
        return version;
    }
    public String getCreator() {
        return creator;
    }
    public void setCreator(String creator) {
        this.creator = creator;
    }
    public Metadata getMetadata() {
        return metadata;
    }
    public void setMetadata(Metadata metadata) {
        this.metadata = metadata;
    }
    public java.util.List<Wpt> getWpts() {
        return wpts;
    }
    public void setWpts(java.util.List<Wpt> wpts) {
        this.wpts = wpts;
    }
    public void addWpt(Wpt wpt) {
        if (this.wpts == null)
            this.wpts = new java.util.ArrayList<>();
        this.wpts.add(wpt);
    }
    public java.util.List<Rte> getRtes() {
        return rtes;
    }
    public void setRtes(java.util.List<Rte> rtes) {
        this.rtes = rtes;
    }
    public void addRte(Rte rte) {
        if (this.rtes == null)
            this.rtes = new java.util.ArrayList<>();
        this.rtes.add(rte);
    }
    public java.util.List<Trk> getTrks() {
        return trks;
    }
    public void setTrks(java.util.List<Trk> trks) {
        this.trks = trks;
    }
    public void addTrk(Trk trk) {
        if (this.trks == null)
            this.trks = new java.util.ArrayList<>();
        this.trks.add(trk);
    }
    @JsonAnyGetter
    public java.util.Map<String,Object> getExtensions() {
        return extensions;
    }
    public void setExtensions(java.util.Map<String,Object> extensions) {
        this.extensions = extensions;
    }
    @JsonAnySetter
    public void putExtensions(String key, Object value) {
        this.extensions.put(key, value);
    }
}

File: kotlin-gpx/src/main/kotlin-gen/Gpx.kt
package com.topografix.gpx

import com.fasterxml.jackson.annotation.JsonAnyGetter
import com.fasterxml.jackson.annotation.JsonAnySetter
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonValue
import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlElementWrapper
import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlProperty
import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlText

enum class FixTypeEnum(@JsonValue val value:String)
{
    _2d("2d"),
    _3d("3d"),
    Dgps("dgps"),
    None("none"),
    Pps("pps")
}

@JsonIgnoreProperties(value = ["schemaLocation"])
data class Gpx(
    @JacksonXmlProperty(localName="version",isAttribute = true) val version:String = "1.1", 
    @JacksonXmlProperty(localName="creator",isAttribute = true) var creator:String = "", 
    var metadata:Metadata? = null, 
    @JacksonXmlElementWrapper(localName="wpts", useWrapping=false) @JacksonXmlProperty(localName="wpt") var wpts:MutableList<Wpt>? = mutableListOf(), 
    @JacksonXmlElementWrapper(localName="rtes", useWrapping=false) @JacksonXmlProperty(localName="rte") var rtes:MutableList<Rte>? = mutableListOf(), 
    @JacksonXmlElementWrapper(localName="trks", useWrapping=false) @JacksonXmlProperty(localName="trk") var trks:MutableList<Trk>? = mutableListOf(), 
    var extensions:Extensions? = null
)
{
}

data class Metadata(
    var name:String? = null, 
    var desc:String? = null, 
    var author:Person? = null, 
    var copyright:Copyright? = null, 
    @JacksonXmlElementWrapper(localName="links", useWrapping=false) @JacksonXmlProperty(localName="link") var links:MutableList<Link>? = mutableListOf(), 
    var time:java.time.LocalDateTime? = java.time.LocalDateTime.now(), 
    var keywords:String? = null, 
    var bounds:Bounds? = null, 
    var extensions:Extensions? = null
)
{
}

data class Wpt(
    @JacksonXmlProperty(localName="lat",isAttribute = true) var lat:Double = 0.0, 
    @JacksonXmlProperty(localName="lon",isAttribute = true) var lon:Double = 0.0, 
    var ele:Double? = 0.0, 
    var time:java.time.LocalDateTime? = java.time.LocalDateTime.now(), 
    var magvar:Double? = 0.0, 
    var geoidheight:Double? = 0.0, 
    var name:String? = null, 
    var cmt:String? = null, 
    var desc:String? = null, 
    var src:String? = null, 
    @JacksonXmlElementWrapper(localName="links", useWrapping=false) @JacksonXmlProperty(localName="link") var links:MutableList<Link>? = mutableListOf(), 
    var sym:String? = null, 
    var type:String? = null, 
    var fix:FixTypeEnum? = null, 
    var sat:Int? = 0, 
    var hdop:Double? = 0.0, 
    var vdop:Double? = 0.0, 
    var pdop:Double? = 0.0, 
    var ageofdgpsdata:Double? = 0.0, 
    var dgpsid:Int? = 0, 
    var extensions:Extensions? = null
)
{
}

data class Rte(
    var name:String? = null, 
    var cmt:String? = null, 
    var desc:String? = null, 
    var src:String? = null, 
    @JacksonXmlElementWrapper(localName="links", useWrapping=false) @JacksonXmlProperty(localName="link") var links:MutableList<Link>? = mutableListOf(), 
    var number:Int? = 0, 
    var type:String? = null, 
    var extensions:Extensions? = null, 
    @JacksonXmlElementWrapper(localName="rtepts", useWrapping=false) @JacksonXmlProperty(localName="rtept") var rtepts:MutableList<Wpt>? = mutableListOf()
)
{
}

data class Trk(
    var name:String? = null, 
    var cmt:String? = null, 
    var desc:String? = null, 
    var src:String? = null, 
    @JacksonXmlElementWrapper(localName="links", useWrapping=false) @JacksonXmlProperty(localName="link") var links:MutableList<Link>? = mutableListOf(), 
    var number:Int? = 0, 
    var type:String? = null, 
    var extensions:Extensions? = null, 
    @JacksonXmlElementWrapper(localName="trksegs", useWrapping=false) @JacksonXmlProperty(localName="trkseg") var trksegs:MutableList<Trkseg>? = mutableListOf()
)
{
}

data class Trkseg(
    @JacksonXmlElementWrapper(localName="trkpts", useWrapping=false) @JacksonXmlProperty(localName="trkpt") var trkpts:MutableList<Wpt>? = mutableListOf(), 
    var extensions:Extensions? = null
)
{
}

data class Copyright(
    @JacksonXmlProperty(localName="author",isAttribute = true) var author:String = "", 
    var year:Int? = 0, 
    var license:String? = null
)
{
}

data class Link(
    @JacksonXmlProperty(localName="href",isAttribute = true) var href:String = "", 
    var text:String? = null, 
    var type:String? = null
)
{
}

data class Email(
    @JacksonXmlProperty(localName="id",isAttribute = true) var id:String = "", 
    @JacksonXmlProperty(localName="domain",isAttribute = true) var domain:String = ""
)
{
}

data class Person(
    var name:String? = null, 
    var email:Email? = null, 
    var link:Link? = null
)
{
}

data class Pt(
    @JacksonXmlProperty(localName="lat",isAttribute = true) var lat:Double = 0.0, 
    @JacksonXmlProperty(localName="lon",isAttribute = true) var lon:Double = 0.0, 
    var ele:Double? = 0.0, 
    var time:java.time.LocalDateTime? = java.time.LocalDateTime.now()
)
{
}

data class Bounds(
    @JacksonXmlProperty(localName="minlat",isAttribute = true) var minlat:Double = 0.0, 
    @JacksonXmlProperty(localName="minlon",isAttribute = true) var minlon:Double = 0.0, 
    @JacksonXmlProperty(localName="maxlat",isAttribute = true) var maxlat:Double = 0.0, 
    @JacksonXmlProperty(localName="maxlon",isAttribute = true) var maxlon:Double = 0.0
)
{
}

class Extensions(
    @JsonIgnore var map:LinkedHashMap<String,String> = linkedMapOf()
)
{
    @JsonAnySetter fun set(key:String, value:String) = this.map.put(key, value)
    @JsonAnyGetter fun all(): LinkedHashMap<String,String> = this.map
    override fun hashCode() = map.hashCode()
    override fun equals(other:Any?) = when {
        this === other -> true
        other is Extensions -> map.size == other.map.size
            && map.all { (k,v) -> v.equals(other.map[k]) }
        else -> false
    }
    override fun toString() = map.toString()
}
